compat/socket.cc

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (C) 1996-2023 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+#include "compat/socket.h"
+
+#if _SQUID_WINDOWS_ || _SQUID_MINGW_
+#include <map>
+
+static_assert(SOCKET_ERROR == -1);
+
+/**
+ * Squid socket code is written to handle POSIX errno codes.
+ * Set errno to the relevant POSIX or WSA code.
+ */ 
+static void
+SetErrnoFromWsaError()
+{
+    // POSIX codes which socket API users may care about
+    static const std::map<int, int> CodeMap = {
+
+        // values checked for by accept(2) callers
+        { WSAECONNABORTED, ECONNABORTED },
+
+        // values checked for by connect(2) callers
+        { WSAEINPROGRESS, EINPROGRESS },
+        { WSAEAFNOSUPPORT, EAFNOSUPPORT },
+        { WSAEINVAL, EINVAL },
+        { WSAEISCONN, EISCONN },
+
+        // values checked by ignoreErrno()
+        { WSAEWOULDBLOCK, EWOULDBLOCK },
+        // WSAEAGAIN no defined
+        { WSAEALREADY, EALREADY },
+        { WSAEINTR, EINTR },
+        // WSARESTART not defined
+
+        // values checked by limitError()
+        { WSAEMFILE, EMFILE },
+        // WSAENFILE not defined
+        
+        // values checked by TunnelStateData::Connection::debugLevelForError()
```

Comment:
I suggest not adding this and all other "checked [for] by ..." comments because they are or will become stale/incorrect as callers change. If some code is using EFOO, add EFOO to this container. No EFOO-specific comment is needed for that entry.

Reply:
These comments are the only documentation we have for why these codes are converted and not all the others. They exist to aid us in future auditing whether they need more additions or removals.